File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char ( toUpper, isLetter, isSpace )

responseFor :: String -> String
responseFor xs = 
  if allSpaces xs then "Fine. Be that way!"
  else if (isYelling xs) && (isQuestion xs) then "Calm down, I know what I'm doing!"
  else if isYelling xs then "Whoa, chill out!"
  else if isQuestion $ trim xs then "Sure."
  else "Whatever."

isQuestion :: String -> Bool
isQuestion str = (last str) == '?'

normalize :: String -> String
normalize str = filter isLetter str    

allSpaces :: String -> Bool
allSpaces str = all isSpace str

isYelling :: String -> Bool
isYelling str = (normalize str /= "") && (map toUpper str) == str

-- from https://stackoverflow.com/a/6270337
trim :: String -> String
trim = f . f
   where f = reverse . dropWhile isSpace
  

File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      String.ends_with?(input,"?") && String.match?(input, ~r/^[^a-z0-9:)]*$/) ->
        "Calm down, I know what I'm doing!"
      String.trim(input) |> String.ends_with?("?") ->
        "Sure."
      String.split(input) == [] ->
        "Fine. Be that way!"
      String.match?(input, ~r/^[,0-9 ]+$/) ->
        "Whatever."
      String.match?(input, ~r/^[^a-z]*$/) ->
        "Whoa, chill out!"
      true ->
        "Whatever."
    end
  end
end
